### PR TITLE
Added Hasu TMK Controller to Via Configurator. 

### PR DESCRIPTION
**Pull request by lbibass**
## Description

made some changes to the matrix.c file, with help from @drashna. I also made the changes required by QMK to add dynamic keymap support from the guide by @nooges. Basically, VIA works with the HHKB now!
## Types of changes

- [ ] Core
- [ ] Bugfix
- [ ] New Feature
- [x] Enhancement/Optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/Layout/Userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by this PR

* 

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document. (https://docs.qmk.fm/#/contributing)
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by mechmerlin** [comment]
I call upon the @Wilba6582 for $approval and other similar $reasons.

**Comment by lbibass** [comment]
@mechmerlin sounds good! I will add it in asap.

**Comment by Wilba6582** [comment]
I think _maybe_ it would be a good idea to make new VID/PID values, at least the current VID `0xFEED` is heavily overused. Perhaps take the whole block of USB device descriptors and put them in `via/config.h` and change `PRODUCT` to be something specific? Actual hardware name etc. Or change it in `hhkb/config.h`. IDK... just an idea.

Also, I'd prefer explicitly turning off features in `via/rules.mk` to ensure keycodes are consistent, copy them from `wilba_tech/*/via/rules.mk`

It's late here so maybe my brain isn't working well, but aren't some of those `#define` in `via/config.h` redefinitions causing warnings or errors? Or does QMK have those suppressed?

**Comment by Wilba6582** [comment]
Expanding on the last point, this could be something I might do in future, to use USB descriptor `PRODUCT` to be something like `KeyboardName (VIA)` to differentiate the VIA build from the non-VIA build. That's not so important for other boards, but kind of nice for something like this so you can see what it's using in the OS Devices page and distinguishing between "default" QMK and VIA... again, just a late night brainfart...

**Comment by drashna** [comment]
> Expanding on the last point, this could be something I might do in future, to use USB descriptor `PRODUCT` to be something like `KeyboardName (VIA)` to differentiate the VIA build from the non-VIA build. That's not so important for other boards, but kind of nice for something like this so you can see what it's using in the OS Devices page and distinguishing between "default" QMK and VIA... again, just a late night brainfart...

That's a really good idea.  


As for the VID/PID thing, yeah.  Though, I think we need a good, official solution here.  What we're doing now "works", but is mostly offsetting tech debt, I think.

>It's late here so maybe my brain isn't working well, but aren't some of those `#define` in `via/config.h` redefinitions causing warnings or errors? Or does QMK have those suppressed?

If they're the same value, then it's fine, actually.  The compiler doesn't mind that.  But if they're different, it will complain.

**Comment by nooges** [comment]
On top of the VID/PID issue, I'm thinking we're going to need a different combo to support HHKB JP as well.

**Comment by lbibass** [comment]
@nooges wouldn't it be possible to read the matrix from the keyboard? Or read the keymap and change which matrix/graphical representation of the keymap is displayed in VIA?

File: keyboards/hhkb/keymaps/via/keymap.c [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+     */
+
+    [HHKB] = LAYOUT(
+        KC_PWR, KC_F1, KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8, KC_F9, KC_F10, KC_F11, KC_F12, KC_INS, KC_DEL,
+        KC_CAPS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_PSCR, KC_SLCK, KC_PAUS, KC_UP, KC_TRNS, KC_BSPC,
+        KC_TRNS, KC_VOLD, KC_VOLU, KC_MUTE, KC_TRNS, KC_TRNS, KC_PAST, KC_PSLS, KC_HOME, KC_PGUP, KC_LEFT, KC_RGHT, KC_PENT,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_PPLS, KC_PMNS, KC_END, KC_PGDN, KC_DOWN, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS)};
+
+const macro_t *action_get_macro(keyrecord_t *record, uint8_t id, uint8_t opt)
```
Comment (drashna) [inline keyboards/hhkb/keymaps/via/keymap.c]: Could you remove the `action_get_macro` function?  It's been deprecated in favor of `process_record_user`.

**Comment by MxBlu** [comment]
> @nooges wouldn't it be possible to read the matrix from the keyboard? Or read the keymap and change which matrix/graphical representation of the keymap is displayed in VIA?

See https://github.com/olivia/via-config/blob/master/app/utils/kle-parser.js
All the layouts are hardcoded, and displayed based on VID/PID, the easiest way to an end is to do as nooges suggests, maybe going forward, it would be nice to implement something more dynamic, provided EEPROM sizes permit.

**Comment by drashna** [comment]
@lbibass if you a chance to update this?

**Comment by lbibass** [comment]
@drashna I added a short readme! Sorry for the delay.

**Comment by drashna** [comment]
I'm sorry to do this to you, but it's bad timing, basically: 


Just a heads up. VIA support is officially "on hold" until such a time that the VIA Dev Team says otherwise.  As such, this PR will be marked as "on hold" until we get their go ahead.

Specifically, they want to be able to focus on generalizing VIA and needed code changes in QMK Firmware so that any board can use VIA.  However, this means putting a pause on adding any new boards, since there will likely be breaking changes and frees up more time to work on these improvements.  So in an effort to make this easier on everyone, we're not accepting any PRs that add VIA support. 

You can read more about this in the discord post, here: https://github.com/olivia/via-config/blob/master/README.md

**Comment by lbibass** [comment]
Okay, sounds good! I'll take a look at it again once the required changes are made.

**Comment by zvecr** [comment]
This can be updated now with the instructions at <https://caniusevia.com/>

**Pull request closed without merging**